examples/stable_diffusion_slackbot/README.md

```diff
@@ -0,0 +1,469 @@
+```
+
+**5. Put it all together…**
+
+Note: the `oxen_row` construct below is a python dictionary matching the schema of our empty CSV stored at `annotations/train.csv` in our oxen repo.
+
+```python
+USER_HASH_LENGTH = 6
+MODEL_VERSION = "0-1"
+def handle_reaction(reaction_data, message_data):
+    file_id = message_data["files"][0]["id"]
+    file_hash = hash_file_id(file_id)
+    prompt = unidecode(message_data["files"][0]["title"])
+    oxen_row = {
+        "prompt": prompt,
+        "path": f"{IMAGE_DIR}/{file_hash}.png",
+        "rating": VOTES[reaction_data['event']['reaction']],
+        "rater": hashlib.sha256(reaction_data['event']['user'].encode("utf-8")).hexdigest()[0:USER_HASH_LENGTH],
+        "model_version": MODEL_VERSION,
+    }
+    commit_df_to_oxen(oxen_row)
+
+# ... back to our main POST route ...
+def post():
+	if data['event']['type'] == 'file_upload':
+		# ... code from earlier omitted ... 
+
+	if data['event']['type'] == 'reaction_added':
+	    if not is_valid_reaction(data):
+	        return "Skipping"
+	    
+	    conversation_id = data['event']['item']['channel']
+	    message_ts = data['event']['item']['ts']
+	    message_data = fetch_message(conversation_id, message_ts)
+	    
+	    if not is_valid_reaction_message(message_data):
+	        return "Skipping"
+	    
+	    thr = Thread(target=handle_reaction, args=[data, message_data["messages"][0]])
+	    thr.start() 
+	    
+	    return "Success"
+	return "Success"
+```
+
+## Results
+
+Let’s see this thing in action! 
+
+Our fine-tuned ox generator is quite good at making cute cartoonish oxen in our target style, but still prone to some rather catastrophic errors, as we saw when asking the bot for an image of our ox *eating* broccoli, not slowly transforming into it.
+
+![A frightening image of our Slackbot, when prompted to generate "an image of the oxenai ox eating brocolli", instead rendering a hybrid image of a creature with the body of an ox and the head of a stalk of broccoli](images/broccoli.png)
+
+To work towards mitigating this, we can pile on the downvotes and have that preference data automatically pushed to our Oxen repo for future fine-tuning.
+
+Our image has made it in, at a hashed path in the `images` folder:
+
+![The broccoli image rendered in the OxenHub UI](images/broccoli-browser.png)
+
+…and our new training observations perfectly reference it from our `train.csv` folder.
+
+![Two "disapprove" rows rendered in the OxenHub UI to reflect our team's opinion of the broccoli image](images/broccoli-tabular.png)
+
+Over the next few months, we’ll be using this bot to generate lots of additional training examples from the comfort of our Slack workspace. With this additional training data generated directly from human feedback, we can work on fine-tuning out some of the glitchier oxen (including the broccoli-ox hybrid shown above) and move towards a more consistent, adaptable ox character.
```

Comment:
Let's actually do this with stable diffusion XL, a modal flask server, and a slack cron job that posts into the channel every morning. I think we can do a similar thing as our github stars repo and generate "random" prompts and then we can select the best generations.